src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -167,21 +173,185 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
+                            {
+                                applicationSettings.Remove(change.Key);
+                            }
+                            else if (change.ChangeType == KeyValueChangeType.Modified)
+                            {
+                                applicationSettings[change.Key] = change.Current;
+                            }
+
+                            // Invalidate the cached Key Vault secret (if any) for this ConfigurationSetting
+                            foreach (IKeyValueAdapter adapter in _options.Adapters)
+                            {
+                                adapter.InvalidateCache(change.Current);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Invalidate all the cached KeyVault secrets
+                        foreach (IKeyValueAdapter adapter in _options.Adapters)
+                        {
+                            adapter.InvalidateCache();
+                        }
+
+                        // Update the cache expiration time for all refresh registered settings and feature flags
+                        foreach (KeyValueWatcher changeWatcher in _options.ChangeWatchers.Concat(_options.MultiKeyWatchers))
+                        {
+                            UpdateCacheExpirationTime(changeWatcher);
+                        }
+                    }
+
+                    if (_options.Adapters.Any(adapter => adapter.NeedsRefresh()) || changedKeyValuesCollection?.Any() == true || keyValueChanges.Any())
+                    {
+                        _applicationSettings = applicationSettings;
+
+                        foreach (KeyValuePair<KeyValueWatcher, KeyValueChange> kvp in keyValueChanges)
+                        {
+                            KeyValueChange keyValueChange = kvp.Value;
+                            KeyValueWatcher changeWatcher = kvp.Key;
+                            KeyValueIdentifier kvIdentifier = new KeyValueIdentifier(changeWatcher.Key, changeWatcher.Label);
+
+                            if (keyValueChange.ChangeType == KeyValueChangeType.Modified)
+                            {
+                                _watchedSettings[kvIdentifier] = keyValueChange.Current;
+                            }
+                            else if (keyValueChange.ChangeType == KeyValueChangeType.Deleted)
+                            {
+                                _watchedSettings.Remove(kvIdentifier);
+                            }
+
+                            // Already updated cache expiration time if refreshAll is true.
+                            if (!refreshAll)
+                            {
+                                UpdateCacheExpirationTime(kvp.Key);
+                            }
+                        }
+
+                        SetData(await PrepareData(applicationSettings, cancellationToken).ConfigureAwait(false));

```

Comment:
It will be good to have some comments here to explain why we do this at the last. I bet I will forget the details 3 months later :)

Reply:
Done!